src/main/scala/io/iohk/ethereum/network/ConnectedPeers.scala

```diff
@@ -0,0 +1,72 @@
+package io.iohk.ethereum.network
+
+import java.net.InetSocketAddress
+
+import akka.actor.ActorRef
+import akka.util.ByteString
+
+case class ConnectedPeers(
+    private val incomingPendingPeers: Map[PeerId, Peer],
+    private val outgoingPendingPeers: Map[PeerId, Peer],
+    private val handshakedPeers: Map[PeerId, Peer]
+) {
+
+  // FIXME: Kept only for compatibility purposes, should eventually be removed
+  lazy val peers: Map[PeerId, Peer] = outgoingPendingPeers ++ handshakedPeers
+
+  private lazy val allPeers: Map[PeerId, Peer] = outgoingPendingPeers ++ handshakedPeers ++ incomingPendingPeers
+
+  def isConnectionHandled(remoteAddress: InetSocketAddress): Boolean =
+    allPeers.values.map(_.remoteAddress).toSet.contains(remoteAddress)
+
+  /*
+      We have the node id of our outgoing pending peers so we could use that in our checks, by rejecting a peer that
+      handshaked to us with the same node id.
+      However, with checking the node id of only handshaked peers we prioritize handshaked peers over pending ones,
+      in the above mentioned case the repeated pending peer connection will eventually die out
+   */
+  def hasHandshakedWith(nodeId: ByteString): Boolean =
```

Comment:
shouldn't we have `lazy val `with all handshaked peer id to avoid this traversing every time ?